src/Datadog.Trace.ClrProfiler.Managed/Emit/ModuleLookup.cs

```diff
@@ -0,0 +1,62 @@
+using System;
+using System.Collections.Concurrent;
+using System.Reflection;
+using System.Threading;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.ClrProfiler.Emit
+{
+    internal static class ModuleLookup
+    {
+        private static readonly ILog Log = LogProvider.GetLogger(typeof(ModuleLookup));
+
+        private static ManualResetEventSlim _populationResetEvent = new ManualResetEventSlim(initialState: true);
+        private static ConcurrentDictionary<Guid, Module> _modules = new ConcurrentDictionary<Guid, Module>();
+
+        public static Module Get(Guid moduleVersionId)
+        {
+            Module value = null;
+
+            // Block if a population event is happening
+            _populationResetEvent.Wait();
```

Comment:
Is it worth adding another `if (_modules.TryGetValue(moduleVersionId, out value)) { return value; }` _before_ the wait? In case it's already cached we can just return right away?

Reply:
In that case, could the `_populationResetEvent.Wait();` just be moved before the `_populationResetEvent.Reset();`?

Reply:
The idea is to prevent as many iterations on the Modules as possible.
I'm okay with adding some logic to not block if it doesn't need to. A second lookup is better than iterating everything.